torchtune/modules/moe/indices.py

```diff
@@ -0,0 +1,353 @@
+        experts_per_rank,
+        num_ranks,
+        BLOCK_SIZE=block_size,
+    )
+    return permuted_indices
+
+
+# reference
+def fill_indices_cpu(
+    tokens_per_expert_group: torch.Tensor,
+    start_index_values: torch.Tensor,
+    write_offsets: torch.Tensor,
+    total_tokens_per_expert: torch.Tensor,
+    experts_per_rank: int,
+    num_ranks: int,
+    max_len: int,
+):
+    # We need to preallocate the output - we ignore device and force it on cpu
+    # device = tokens_per_expert_group.device
+    permuted_indices = torch.full(
+        (max_len,),
+        -1,
+        dtype=torch.int32,
+    )  # device=device)
+    # Fill the permuted indices
+    # For each local expert
+    for e in range(experts_per_rank):
+        write_start = write_offsets[e].item()
+        total_tokens = total_tokens_per_expert[e].item()
+
+        # For each remote rank
+        # we can skip this expert if it has no tokens, already filled with -1
+        if total_tokens > 0:
+            for r in range(num_ranks):
+                i = r * experts_per_rank + e
+                start_index = start_index_values[i].item()
+                length = tokens_per_expert_group[i].item()
+                # Fill in the indices
+                if length > 0:
+                    end_idx = min(write_start + length, max_len)
+                    permuted_indices[write_start:end_idx] = torch.arange(
+                        start_index,
+                        start_index + (end_idx - write_start),
+                        dtype=torch.int32,
+                        # device=device,
+                    )
+                write_start += length
+    return permuted_indices
+
+
+def generate_permute_indices(
+    tokens_per_expert_group: torch.Tensor,
+    experts_per_rank: int,
+    num_ranks: int,
+    max_len: int,
+    alignment: int,
+    use_cpu: bool = False,
+):
+    """
+    Prepare permutation indices and the number of tokens for each expert.
+
+    Args:
+        tokens_per_expert_group (torch.Tensor): number of tokens for each expert from all ranks.
+        experts_per_rank (int): number of experts per rank.
+        num_ranks (int): number of ranks.
+        max_len (int): maximum length of the output index vector.
+        alignment (int): alignment for each returned element in `m_sizes` and padding min for zero token experts.
+        use_cpu (bool): whether to use CPU implementation.
+
+
+    Returns:
+        permuted_indices: Tensor of indices that map original token order to the expert-grouped order.
+        m_sizes: aligned number of tokens for each expert (padded to alignment boundary).
+        m_offsets: Cumulative sum of m_sizes. The exclusive ending position for each expert's tokens.
+
+    Explanatory details:
+        `tokens_per_expert_group` is of shape (num_ranks * experts_per_rank,), for example:
+        From: |       rank 0      |       rank 1      |
+        To:   | E0 | E1 | E2 | E3 | E0 | E1 | E2 | E3 |
```

Comment:
Noob q on this diagram: does this mean that we actually have the same expert on multiple ranks (i.e. there are only 4 experts total)? I ask because the arg `experts_per_rank` seems to imply that we actually have different experts on different ranks (I know it's moot for the implementation in its current form anyways so not a huge deal).

Reply:
In our case each rank runs all the experts, but all mms happen in parallel. 
So experts_per_rank is always 16 (all experts).

So we are not bucketing experts by ranks. I think this logic will be similar, to what TorchRec does with embeddings, basically each expert works as a lookup Table.